<a href="https://colab.research.google.com/github/haizznaam/Content-Based-and-Tensorflow-Movie-Recommender-System/blob/main/surprise_knn_recommendation_sys.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [2]:
import os
os.chdir('/content/drive/MyDrive/Movie Recommendation System/The-Movie-Dataset/')
print("We are currently in the folder of ", os.getcwd())

We are currently in the folder of  /content/drive/MyDrive/Movie Recommendation System/The-Movie-Dataset


In [5]:
pip install surprise

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 772.0/772.0 kB 4.0 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
  Created wheel for scikit-surprise: filename=scikit_surprise-1.1.3-cp310-cp310-linux_x86_64.whl size=3162995 sha256=2ebc73162473f1fdbb3212a6af87745e1f73817b07b424f5ade6095f03399ad9
  Stored in directory: /root/.cache/pip/wheels/a5/ca/a8/4e28def53797fdc4363ca4af740db15a9c2f1595ebc51fb445
Successfully built scikit-surprise


In [6]:
import numpy as np
import pandas as pd
from surprise import Dataset, Reader
from surprise.model_selection import train_test_split, GridSearchCV
from surprise import SVD, KNNBasic, NMF, SlopeOne, CoClustering, BaselineOnly
from surprise import accuracy

**Collaborative filtering recommends items based on the preferences of users or items similar to the target**. There are two main types:

  >**User-based**: Recommends items liked by users >with similar preferences.
  >
  >**Item-based**: Recommends items similar to those > the user has liked or interacted with.

Both methods rely on the collective behavior of users to make personalized recommendations

In [7]:
data = pd.read_csv('ratings_small.csv')
df = pd.DataFrame(data)
df.head()

userId  movieId  rating   timestamp
0       1       31     2.5  1260759144
1       1     1029     3.0  1260759179
2       1     1061     3.0  1260759182
3       1     1129     2.0  1260759185
4       1     1172     4.0  1260759205

In [8]:
df = df.drop(columns = 'timestamp')
df.head(3)

userId  movieId  rating
0       1       31     2.5
1       1     1029     3.0
2       1     1061     3.0

In [10]:
df.describe()

userId        movieId         rating
count  100004.000000  100004.000000  100004.000000
mean      347.011310   12548.664363       3.543608
std       195.163838   26369.198969       1.058064
min         1.000000       1.000000       0.500000
25%       182.000000    1028.000000       3.000000
50%       367.000000    2406.500000       4.000000
75%       520.000000    5418.000000       4.000000
max       671.000000  163949.000000       5.000000

In [11]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100004 entries, 0 to 100003
Data columns (total 3 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   userId   100004 non-null  int64  
 1   movieId  100004 non-null  int64  
 2   rating   100004 non-null  float64
dtypes: float64(1), int64(2)
memory usage: 2.3 MB


In [12]:
df.isna().sum()

userId     0
movieId    0
rating     0
dtype: int64

**Reading the data with suprise Libarary**

In [13]:
reader = Reader(rating_scale=(0,5))
data = Dataset.load_from_df(df , reader)

**Making a train and test data for evaluation**

In [16]:
train_set , test_set = train_test_split(data ,test_size = 0.2, random_state=42)

**I am going to use these algorithms on the data to see which works better**

In [15]:
algorithms = [SVD(), KNNBasic(), NMF(), SlopeOne(), CoClustering(), BaselineOnly()]

**Making a loop to test the algorithms**

In [24]:
knn = KNNBasic()
knn.fit(train_set)

print("=" * 40)

test_predictions = knn.test(test_set)
test_rmse = accuracy.rmse(test_predictions)
test_mae = accuracy.mae(test_predictions)
print("KNN Basic fillna with 0")
print('Test RMSE: ', test_rmse)
print('Test MAE: ', test_mae)

Computing the msd similarity matrix...
Done computing similarity matrix.
RMSE: 0.9672
MAE:  0.7456
KNN Basic fillna with 0
Test RMSE:  0.9671524784958385
Test MAE:  0.7455556591578778


In [25]:
from surprise import KNNWithMeans
knnmean = KNNWithMeans()
print("=" * 40)
knnmean.fit(train_set)
test_predictions = knnmean.test(test_set)
test_rmse = accuracy.rmse(test_predictions)
test_mae = accuracy.mae(test_predictions)
print("KNN with Means")
print('Test RMSE: ', test_rmse)
print('Test MAE: ', test_mae)

Computing the msd similarity matrix...
Done computing similarity matrix.
RMSE: 0.9234
MAE:  0.7075
KNN with Means
Test RMSE:  0.9233587733995928
Test MAE:  0.7075044755093767


| Model                     | Test RMSE           | Test MAE            |
|---------------------------|---------------------|---------------------|
| **KNN Basic fillna with 0** | 0.967  | 0.746  |
| **KNN with Means**        | 0.923  | 0.707  |
